### ISSUE TITLE
Images don't render - raise javascript error + performance discussion

### ISSUE THREAD

**Issue by ianhi** [vispy/vispy.js#31]
I would like to display an image in a jupyter notebook but this code raises a javascript error:

```python
import numpy as np
import vispy
import vispy.scene
from vispy import app
from vispy.scene import visuals
print("VISPY VERSION:", vispy.__version__)
from vispy.io import load_data_file, read_png


# Define a function to tranform a picture to gray
def rgb2gray(rgb):
    return np.dot(rgb[...,:3], [0.299, 0.587, 0.114])
app.use_app('ipynb_webgl')

ca = vispy.scene.SceneCanvas(keys='interactive', show=True)
ca.show()
view = ca.central_widget.add_view()
img_data = read_png(load_data_file('mona_lisa/mona_lisa_sm.png'))

# Apply transformation
img_data = rgb2gray(img_data)

image = visuals.Image(img_data, cmap='grays')#, parent=view.scene)


color = np.array([
    [255, 0, 0]
])
view.add(image)

view.camera = 'turntable' # or try 'arcball'

axis = visuals.XYZAxis(parent=view.scene)
app.run()
```
Results in this in the console:
![image](https://user-images.githubusercontent.com/10111092/87336481-ebd4e200-c50f-11ea-8a74-c1de1aa21df7.png)
and an all black output

## Context
Jupyterlab: 2.1.5
Vispy: Installed via git today 0.6.5.dev174+gbb2cd96e
vispy.js: 0.3.0
Browser: Firefox 78
OS: PopOS 20.04 (ubuntu derivative)

Any help much appreciated thanks!

**Comment by djhoese** [comment]
Do any of the notebook examples work for you? I wonder if some of the recent changes to the ImageVisual (by @tlambert03 in https://github.com/vispy/vispy/pull/1844) are breaking WebGL compatibility. However, that seems extremely unlikely for the `max` function that it is complaining about.

**Comment by djhoese** [comment]
For example: the rotating cube example https://github.com/vispy/vispy/tree/master/examples/ipynb

**Comment by ianhi** [comment]
oooh I hadn't seen those examples. I was taking things from SO and issues involving jupyter.

All of the example notebooks work for me. As well both of these examples work:
```python
import vispy.app
vispy.app.use_app('ipynb_webgl')
from vispy import plot as vp
fig = vp.Fig(show=False)
fig1 = fig[0, 0]
fig1.plot(range(10000),marker_size=0)
fig.show(run=True)
```

and this example on which my image case is partially based:
```python
import numpy as np
import vispy
import vispy.scene
from vispy import app
from vispy.scene import visuals
print("VISPY VERSION:", vispy.__version__)
app.use_app('ipynb_webgl')

ca = vispy.scene.SceneCanvas(keys='interactive', show=True)
ca.show()
view = ca.central_widget.add_view()
scatter = visuals.Markers()


color = np.array([
    [255, 0, 0]
])
data = np.random.random((2048, 3))
scatter.set_data(data, edge_color=None, face_color=color/255.0, size=5)
view.add(scatter)

view.camera = 'turntable' # or try 'arcball'

axis = visuals.XYZAxis(parent=view.scene)
app.run()
```

**Comment by tlambert03** [comment]
> I wonder if some of the recent changes to the ImageVisual (by @tlambert03 in vispy/vispy#1844) are breaking WebGL compatibility.

😬 !  definitely let me know if I can help!

**Comment by ianhi** [comment]
Also I have a more general question about this widgets architecture/performance that I can open a separate issue for or ask elsewhere if you'd like.

Can I expect vispy to be more responsive to user interactionst in a jupyter notebook than for example the matplotlib widget ([ipympl](https://github.com/matplotlib/ipympl))? In the readme you write that you send `glir` over a websocket, are you opening your own websocket or relying on the notebook comms? I think one of the big sources of lag in user interaction in widgets is the message speed between javascript and python. I characterized it in this repo https://github.com/ianhi/widget_message_speed#results So for example when I tried to make an image thresholding widget that would work in a jupyter notebook using matplolib (which has to rely on python callbacks) it was fairly laggy:
![slighlty-less](https://user-images.githubusercontent.com/10111092/87343590-1e380c80-c51b-11ea-9484-9ef0959e8883.gif)

I think this due to:
1. message timing latency
2. matplotlib being slower than vispy to render an image (I think?)


I know that vispy can be wildly faster than matplotlib at least when run outside of a notebook, but is it possible to get the same speedups in the notebook?

**Comment by djhoese** [comment]
## Your original error

I wonder if your code is failing because of the data type of the ImageVisual that is being created. What's weird, if I'm reading the error message correctly, the issue is in the clearColor:

https://github.com/vispy/vispy.js/blob/d17bb62b10db1f5705093c53d589e213d655c8c9/lib/gloo.glir.js#L15

But that doesn't make sense. Anyway, what happens if you remove your `rgb2grays` and `cmap=` parts. Does it handle that?

## Performance

The vispy jupyter extension is considered experimental (I tried to mark it that way in the installation instructions). The main reason for this designation is my opinion of the performance. The widget/extension was written by some of the original vispy maintainers and I updated it where I could to make it work with the new jupyter widget (ipywidget) systems. The original version was so old that notebooks were still ipython notebooks (jupyter didn't exist yet). We are using the jupyter widget comms system to create a web socket. This means that in a normal user interactive visualization you have the following series of events:

1. User click/keyboard event
2. <event sent to python kernel>
3. Python interprets event, updates various properties, generates vispy GLIR commands to send to the client.
4. <GLIR commands sent as individual messages as tuples of strings/numbers)>
5. Javascript interprets the GLIR commands and calls the relevant WebGL functions.

The roundtrip-ness of this is extremely slow just with the communication, but the generation and interpretation of the GLIR commands doesn't help either. I've never profiled the javascript but I'm guessing it isn't as fast as it could be. Even with the rotating cube example, I think you can hit space to pause it and if you leave it running for long enough (<30 seconds even) and pause it you'll see that it takes a decent amount of time for the browser canvas to respond to that event. Besides comms I think this is just the amount of time it takes for the javascript to parse all the GLIR commands coming in.

### Other things to consider

1. Implement a vispy kernel in the browser using web assembly or javascript (if needed). Basically handle user events in the browser as much as possible and limit communication to the python kernel where possible. I have no idea how this would actually be implemented.
2. Skip making WebGL better and focus on WebGPU programming. This is something one of the original vispy maintainers is working on outside the browser and outside the vispy project (a python-interface to a WebGPU library for visualizations).

**Comment by ianhi** [comment]
> Your original error

No change if I remove those. 

> But that doesn't make sense.

I agree. I actually also ended up there when clicking through the FF debugger. That's when I decided I was thoroughly confused and opened this issue. 

> The vispy jupyter extension is considered experimental (I tried to mark it that way in the installation instructions). 

It was clearly marked fear not! I just was being optimistic about the meaning of experimental :)

> Implement a vispy kernel in the browser using web assembly or javascript (if needed). Basically handle user events in the browser as much as possible and limit communication to the python kernel where possible. I have no idea how this would actually be implemented.

For the widgets I've made this was generally the solution. User interaction basically has to happen in the front end if its going to be pleasant. 

> a python-interface to a WebGPU library for visualizations

That sounds cool, do you have a link to this?


## Performance thoughts

I wonder if there is a route around the comms that could be created that would be faster - like would a custom websocket be quicker than whatever the comms are abstracting away.


Another library you may want to look towards for inspiration is [ipycanvas](https://github.com/martinRenou/ipycanvas) and in particular the `hold_canvas` context manager https://ipycanvas.readthedocs.io/en/latest/basic_usage.html#optimizing-drawings

I think that these use interaction problems are generally an issue in widgets so if you ask about them in quantstack gitter channel (gitter channel for jupyter widgets) https://gitter.im/QuantStack/Lobby# those are the people who are the most likely to have helpful advice.

**Comment by ianhi** [comment]
Also, there can be limits on the rate of messages in the comms. See discussion here https://github.com/martinRenou/ipycanvas/issues/102

**Comment by djhoese** [comment]
Best I can think of next is to transition between the example you started with (which you said worked, right?)
 and slowly transition to what you have that doesn't work.

WebGPU: https://github.com/vispy/vispy/issues/1484#issuecomment-651859820

I'm not familiar with the limitations of the jupyter widget comms so I wouldn't be surprised if a separate dedicated socket would help. The `hold_canvas` idea is interesting. We have some similar handling in vispy as far as "blocking" events and updates but that is still on the python side and doesn't build up the events/commands that will be sent.

Why the quantstack gitter versus the https://gitter.im/jupyter-widgets/Lobby gitter?

**Comment by ianhi** [comment]
> Why the quantstack gitter versus the https://gitter.im/jupyter-widgets/Lobby gitter?

In large part because I somehow was unaware of that link :joy:. I've asked multiple widget dev questions in the quantstack gitter and have always received very helpful answers there. It's possible that the quantstack gitter is more dev focused, the jupyter-widgets seems to mostly be regarding questions on usage of existing widgets.

> I'm not familiar with the limitations of the jupyter widget comms so I wouldn't be surprised if a separate dedicated socket would help. 

I just asked about this here: https://gitter.im/QuantStack/Lobby?at=5f0cd1033e4a827d19c40480

**Comment by ianhi** [comment]
> Best I can think of next is to transition between the example you started with (which you said worked, right?)
and slowly transition to what you have that doesn't work.

I can give this a go but unfortunately the last vispy code I wrote was in November 2017 ([fluid sim!](https://github.com/ianhi/courses/blob/6b8958ec9a6cfb8748553b0815e71341d48164dc/AC274/Final%20Project/final_code/visualization.py)) and I think that even that was mostly copied from an example. So I may not have enough knowledge of the api to do this properly. Is there any transition between switching from `Markers` to `Image`?

**Comment by djhoese** [comment]
The VisPy Python library now depends on a new `jupyter_rfb` backend (to be released in 0.8.0, not out yet). As such, we have deprecated and removed the `ipynb_webgl` backend. This means the `vispy.js` project will not be used by the VisPy project anymore and will not receive active development. I'm closing this issue because of this and will be archiving the repository making it read-only.
